### PR TITLE
add privacy preservation demo

### PR DESCRIPTION
**Pull request by dkupnicki**

### CODE REVIEW

**Comment by jan-grzybek-ampere** [comment]
@MarcelWilnicki can you take a look?

**Comment by MarcelWilnicki** [comment]
@dkupnicki could you please explain roughly what this PR is about?

**Comment by dkupnicki** [comment]
It's a demo which takes a video as an input and using object detection and pose estimation models produces a video in which the people are pixelated, so that their privacy is preserved.

File: demos/privacy_preservation/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+## Models
+
+The demo uses two models, one for object detection and one for pose estimation. You can download them here:
+- https://tfhub.dev/tensorflow/lite-model/efficientdet/lite2/detection/default/1
+- https://tfhub.dev/google/lite-model/movenet/singlepose/lightning/3
+
+## How to run
+`OMP_NUM_THREADS=4 AIO_PROCESS_MODE=0 python3 run.py --model-path lite-model_movenet_singlepose_lightning_3.tflite --detection-model-path lite-model_efficientdet_lite2_detection_default_1.tflite`
+
```
Comment (jan-grzybek-ampere) [inline demos/privacy_preservation/README.md:11]: Consider adding a requirements.txt file required to make this thing work in our release docker - that will actually mean just one dep:
flask

Comment (jan-grzybek-ampere) [reply demos/privacy_preservation/README.md:11]: Screenshot 2022-09-09 at 14 11 35
Perhaps having some default welcome image would be nice - like big ampere logo or something

Comment (jan-grzybek-ampere) [reply demos/privacy_preservation/README.md:11]: Your write-up on this demo could be linked somewhere in this read-me.

**Review by jan-grzybek-ampere — COMMENTED** [review]
>     self.blurred, self.pose = self.blur_humans(image)
>   File "/aml/demos/privacy_preservation/postprocessing.py", line 141, in blur_humans
>     bbox = [clamp(round(self.bboxes[idx][0] * npimg.shape[0]), 0, npimg.shape[0]),
> IndexError: index 11 is out of bounds for axis 0 with size 11
> Exception in thread Thread-23:
> Traceback (most recent call last):
>   File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
>     self.run()
>   File "/usr/lib/python3.8/threading.py", line 870, in run
>     self._target(*self._args, **self._kwargs)
>   File "/aml/demos/privacy_preservation/postprocessing.py", line 97, in blur
>     self.blurred, self.pose = self.blur_humans(image)
>   File "/aml/demos/privacy_preservation/postprocessing.py", line 141, in blur_humans
>     bbox = [clamp(round(self.bboxes[idx][0] * npimg.shape[0]), 0, npimg.shape[0]),
> IndexError: index 11 is out of bounds for axis 0 with size 11

I've got this error with the attached video

https://user-images.githubusercontent.com/52271793/189348730-0010b726-1533-47e2-ab2e-385fe515f4ca.mp4

**Review by jan-grzybek-ampere — APPROVED** [review]
I was able to run the demo with my video in the end - issue I saw happens when there's first an attempt to use a webcam (failed in my case), followed by an upload of video file - consider fixing this or adding some warning.

I'm fine with merging this code though - it looks fine and mature enough.
Few things/ideas to consider:

- maybe the video file should be compressed by default to a lower, fixed resolution after being received, before being streamed back? (to increase frame rate) also perhaps re-encoded into some lightweight codec? having fixed parameters of your stream would allow you to make the video smoother, but also in case making it smooth would still not be possible - knowing the approx. latency would allow you to drop some frames to at least make it in-sync with real time
- please make an issue in dlslib repo using your demo as an example of how multi threading is not possible with aio as of today
- video that I've added in one of comments show that some faces close to camera don't get blurred (or don't get blurred in time), especially after disappearing for a moment - Im aware that this can't be fully mitigated, but maybe it can be made better by playing around with thresholds? In case of this demo its better to blur too much than to blur too little. It could also be mitigated by having some "momentum" added (i.e. when there is a face being tracked and blurred, let's calculate its trajectory and keep on blurring the area we would expect it to be at in case it disappears)

File: demos/privacy_preservation/pose.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    def start(self):
+        self.stopped = False
+        Thread(target=self.detect, args=()).start()
+        return self
+
+    def detect(self):
+        while not self.stopped:
+            idx = self.det_pose_queue.get()
+            if idx is None:
+                for i in range(8):
```
Comment (jan-grzybek-ampere) [inline demos/privacy_preservation/pose.py:52]: Worth keeping values like 8 here in some config file or as a global variable - named and perhaps explained.

**Pull request merged**